File: app/src/main/java/com/tim/iot/auth/AuthServer.java
package com.tim.iot.auth;

import android.text.TextUtils;
import com.tim.common.Logger;
import com.tim.iot.auth.entity.WebSocketInfo;
import com.tim.iot.auth.rx.IWebSocket;
import com.tim.iot.device.entity.AccountInfo;
import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.TimeoutException;

/**
 * AuthServer
 *
 * @author Tell.Tim
 * @date 2019/12/24 19:33
 */
public class AuthServer implements IAuthServer {
    private static final String TAG = "AuthServer";
    private static final Logger logger = Logger.getLogger(TAG);
    private IWebSocket webSocket;
    private Disposable disposable;
    public AuthServer() {
        this.webSocket = new Builder().setHeartBeat(true).setShowLog(true, TAG).build();
    }

    @Override
    public synchronized void connect(String url, int timeoutOfSecond,
            IConnectAuthServerCallback connectAuthServerCallback) {
        this.webSocket.connect(url, timeoutOfSecond).subscribe(new Observer<WebSocketInfo>() {
            @Override
            public void onSubscribe(Disposable d) {
                logger.d("订阅关系已建立");
                disposable = d;
            }

            @Override
            public void onNext(WebSocketInfo webSocketInfo) {
                if (webSocketInfo.isConnected()) {
                    logger.d("连接成功");
                    connectAuthServerCallback.onConnectSuccess();
                } else if (webSocketInfo.getSimpleMsg() != null) {
                    String simpleMessage = webSocketInfo.getSimpleMsg();
                    logger.d("收到消息: " + simpleMessage);
                    if (!TextUtils.isEmpty(simpleMessage) && simpleMessage.startsWith("confirm")) {
                        String[] convertArray = simpleMessage.split("#");
                        AccountInfo accountInfo = new AccountInfo();
                        accountInfo.setAccount(convertArray[1]);
                        accountInfo.setCreateAt(Long.valueOf(convertArray[2]));
                        connectAuthServerCallback.onConfirm(accountInfo);
                    }
                } else if (webSocketInfo.getByteStringMsg() != null) {
                    logger.d("receive byte message");
                } else if (webSocketInfo.isReconnect()) {
                    logger.d("正在重连");
                } else if (webSocketInfo.isPreConnect()) {
                    logger.d("准备连接");
                }
            }

            @Override
            public void onError(Throwable e) {
                if (e instanceof TimeoutException) {
                    logger.d("超时未响应");
                    connectAuthServerCallback.onTimeOut();
                } else {
                    logger.d("连接异常 " + e.getMessage());
                    connectAuthServerCallback.onConnectError(new Exception(e));
                }
            }

            @Override
            public void onComplete() {
                logger.d(" --> 处理完毕");
            }
        });
    }

    @Override
    public void closeConnect(String url) {
        logger.d("closeConnect");
        if (!disposable.isDisposed()) {
            logger.d("dispose");
            disposable.dispose();
        } else {
            logger.d("dispose isDisposed");
        }
    }

    @Override
    public void forceClose() {
        this.webSocket.forceClose();
    }
}

File: app/src/main/java/com/tim/iot/auth/AuthServer.kt
package com.tim.iot.auth

import android.text.TextUtils
import com.tim.common.Logger
import com.tim.iot.auth.entity.WebSocketInfo
import com.tim.iot.auth.rx.IWebSocket
import com.tim.iot.device.entity.AccountInfo
import io.reactivex.Observer
import io.reactivex.disposables.Disposable
import java.util.concurrent.TimeoutException

/**
 * AuthServer
 *
 * @author Tell.Tim
 * @date 2019/12/24 19:33
 */
class AuthServer : IAuthServer {
    private val webSocket: IWebSocket
    private var disposable: Disposable? = null

    init {
        this.webSocket = Builder().setHeartBeat(false).setShowLog(true, TAG).build()
    }

    @Synchronized override fun connect(url: String, timeoutOfSecond: Int,
        connectAuthServerCallback: IAuthServer.IConnectAuthServerCallback) {
        this.webSocket.connect(url, timeoutOfSecond).subscribe(object : Observer<WebSocketInfo> {
            override fun onSubscribe(d: Disposable) {
                logger.d("订阅关系已建立")
                disposable = d
            }

            override fun onNext(webSocketInfo: WebSocketInfo) {
                if (webSocketInfo.isConnected) {
                    logger.d("连接成功")
                    connectAuthServerCallback.onConnectSuccess()
                } else if (webSocketInfo.simpleMsg != null) {
                    val simpleMessage = webSocketInfo.simpleMsg
                    logger.d("收到消息: " + simpleMessage!!)
                    if (!TextUtils.isEmpty(simpleMessage) && simpleMessage.startsWith("confirm")) {
                        val convertArray =
                            simpleMessage.split("#".toRegex()).dropLastWhile { it.isEmpty() }
                                    .toTypedArray()
                        val accountInfo = AccountInfo()
                        accountInfo.account = convertArray[1]
                        accountInfo.createAt = java.lang.Long.valueOf(convertArray[2])
                        connectAuthServerCallback.onConfirm(accountInfo)
                    }
                } else if (webSocketInfo.byteStringMsg != null) {
                    logger.d("receive byte message")
                } else if (webSocketInfo.isReconnect) {
                    logger.d("正在重连")
                } else if (webSocketInfo.isPreConnect) {
                    logger.d("准备连接")
                }
            }

            override fun onError(e: Throwable) {
                if (e is TimeoutException) {
                    logger.d("超时未响应")
                    connectAuthServerCallback.onTimeOut()
                } else {
                    logger.d("连接异常 " + e.message)
                    connectAuthServerCallback.onConnectError(Exception(e))
                }
            }

            override fun onComplete() {
                logger.d(" --> 处理完毕")
            }
        })
    }

    override fun closeConnect(url: String) {
        logger.d("closeConnect")
        if (!disposable!!.isDisposed) {
            logger.d("dispose")
            disposable!!.dispose()
        } else {
            logger.d("dispose isDisposed")
        }
    }

    override fun forceClose() {
        this.webSocket.forceClose()
    }

    companion object {
        private val TAG = "AuthServer"
        private val logger = Logger.getLogger(TAG)
    }
}